DeDriftAndResample/DeDriftAndResamplePipeline.sh

```diff
@@ -60,30 +60,34 @@ PARAMETERs are [ ] = optional; < > = user supplied value
        (e.g. \${Subject}.\${Hemisphere}.sphere.\${RegName}.native.surf.gii)
-   [--dedrift-reg-files=</Path/to/File/Left.sphere.surf.gii@/Path/to/File/Right.sphere.surf.gii>]
+  [--dedrift-reg-files=</Path/to/File/Left.sphere.surf.gii@/Path/to/File/Right.sphere.surf.gii>]
        Path to the spheres output by the MSMRemoveGroupDrift pipeline
-   [--concat-reg-name=<regname>] String corresponding to the output name of the dedrifted registration
+  [--concat-reg-name=<regname>] String corresponding to the output name of the dedrifted registration
    --maps=<non@myelin@maps> @-delimited map name strings corresponding to maps that are not myelin maps 
        (e.g. sulc@curvature@corrThickness@thickness)
-   [--myelin-maps=<myelin@maps>] @-delimited map name strings corresponding to myelin maps 
+  [--myelin-maps=<myelin@maps>] @-delimited map name strings corresponding to myelin maps
        (e.g. MyelinMap@SmoothedMyelinMap). No _BC, this will be reapplied.
-   [--multirun-fix-names=<day1run1@day1run2%day2run1@day2run2>] @ and % delimited list of lists of
+  [--multirun-fix-names=<day1run1@day1run2%day2run1@day2run2>] @ and % delimited list of lists of
        fMRIName strings that will have multi-run ICA+FIX reapplied to them (could be either rfMRI or
        tfMRI). Requires specifying --multirun-fix-concat-names also, with same number of concat names as
        lists of runs in this option.
-   [--multirun-fix-concat-names=<day1_concat@day2_concat>] @-delimited list of names of the concatenated
+  [--multirun-fix-concat-names=<day1_concat@day2_concat>] @-delimited list of names of the concatenated
        timeseries, only required when using --multirun-fix-names.
-   [--multirun-fix-extract-names=<day1run1@day1run2%day2run1@day2run2>] @ and % delimited list of lists of
+  [--multirun-fix-extract-names=<day1run1@day1run2%day2run1@day2run2>] @ and % delimited list of lists of
        fMRIName strings to extract, one list for each multi-run ICA+FIX group in --multirun-fix-names (use
        NONE to skip a group), only required when using --multirun-fix-extract-concat-names.  Exists to
        enable extraction of a subset of the runs in a multi-run ICA+FIX group into a new concatenated
        series (which is named using --multirun-fix-extract-concat-names).
-   [--multirun-fix-extract-concat-names=<day1_newconcat@day2_newconcat>] @-delimited list of names for the
+  [--multirun-fix-extract-concat-names=<day1_newconcat@day2_newconcat>] @-delimited list of names for the
        concatenated extracted timeseries, one for each multi-run ICA+FIX group (i.e. name in
        --multirun-fix-concat-names; use NONE to skip a group).
-   [--fix-names=<ICA+FIXed@fMRI@Names>] @-delimited fMRIName strings corresponding to maps that will 
+  [--multirun-fix-extract-extra-regnames=<regname@regname>] extract MR FIX runs for additional surface
+       registrations, often MSMSulc.
```

Comment:
So, if wanting to apply to MSMSulc, do you use `NONE` as the argument?  `MSMSulc` as the argument? Either?

Reply:
MSMSulc.  not specifying it, or using NONE, will cause it to not do any surface registration other than that in the --concat-reg-name argument.

Reply:
Can we make that explicit in the usage, since I believe in other scripts that `NONE` represents `MSMSulc`?

Reply:
Now I'm confused.  The code at one point checks `if [[ "$mrFIXExtractExtraRegNames" == NONE ]]`.  I don't see anywhere that it is checking for `MSMSulc` as the string.

Relatedly, what if someone tried to run MSMAll as the `-extra-regnames` and MSMSulc (`NONE`) as the RegName?  Would that work?

Reply:
Other scripts might use MSMSulc by default (and "NONE" is the old convention for "not specified"), and specifically require one and only one surface registration.  I don't know of a script where specifying "NONE" means "MSMSulc", and specifying "MSMSulc" is wrong.  However, the new Extract... script currently breaks from that convention somewhat, in that currently, specifying "MSMSulc" to it actually is wrong, which perhaps should be changed.

This argument is an optional array argument, and defaults to being empty - the old "NONE" convention to act as though it wasn't specified was added only for consistency with the existing conventions of the script.

What about this small change:
```
  [--multirun-fix-extract-extra-regnames=<regname@regname>] extract MR FIX runs for additional surface
       registrations, e.g. --multirun-fix-extract-extra-regnames=MSMSulc
```

Reply:
Both MSMSulc and NONE should probably do the same thing.

Reply:
Isn't it the other way around -- using `NONE` for RegName implies "MSMSulc"?  That's how the ReApplyFix*.sh scripts are set up...

Reply:
@mharms https://github.com/Washington-University/HCPpipelines/pull/182/files#diff-29e440b6c98c0b8335ae2638db8ece93R970 is the MSMSulc test.

So, it looks like ReApply... would error if you said `MSMSulc`.  That is unfortunate.  I see two possibilities - we can either treat the relevant option as referring to the concept of registration type (building in handling of the naming quirks), or we can use it to refer only to naming convention (specify empty string to use whatever doesn't name its registration type).  Referring only to the naming convention would pose some issues, as specifying an array containing only the empty string would get tricky to distinguish from "not specified".

Either way, `NONE` should really only be used to say "don't use this argument", and only in old scripts that already use that convention (because they didn't handle the empty string properly).  Thus, `NONE` should do something different than `MSMSulc` for this array argument in this script.

Reply:
ok

Reply:
If the concat reg name was something other than MSMAll, and MSMAll was in the `--multirun-fix-extract-extra-regnames` argument, yes, that would currently work (assuming the concat reg name was valid).

Reply:
Can we maybe beef up the usage statement to make clearer the distinction between `--registration-name` and `--concat-reg-name`?

I think we should change the Extract script so that specifying `MSMSulc` is correct.

FWIW, it appears that the TaskAnalysis*.sh, MakeAverageDataset.sh, and RestingStateStats.sh scripts also allow for handling RegName=NONE as MSMSulc.

Reply:
`--registration-name` and `--concat-reg-name` are part of the prior functionality of the script.  I appear to have run my tests specifying `MSMAll` for both, which appears to be the way to disable the dedrifting step.  It looks like the only reason for having this as two arguments is dedrifting, which should **not** be done on subjects' data anyway.

For scripts that require the use of surface data, defaulting to the MSMSulc data when `NONE` specifies "don't use this argument" is tolerable.  That doesn't mean that `NONE` is being interpreted as explicitly specifying MSMSulc.

Reply:
Can we just make that a bit more explicit in the usage notes?  Also, do we actually know if specifying `MSMSulc` for `--registration-name` and `--concat-reg-name` would work?  I suspect that might be a situation that the script was never designed to handle...

Reply:
That is just my guess at the reason for those arguments, and Matt has not always agreed with me about dedrifting.

Given the file naming quirks of MSMSulc, I wouldn't expect it to work as the concatregname/regname, I didn't write that part of the code.  I'm not clear on what point you are making, though.

Reply:
I was just cycling back to my earlier comment:

> Relatedly, what if someone tried to run MSMAll as the -extra-regnames and MSMSulc (NONE) as the RegName? Would that work?

So while using `MSMAll` as the -extra-regnames argument might be fine, using `MSMSulc` for --registration-name and --concat-reg-name is not.

Reply:
Right, I expect MSMSulc to fail as `--concat-reg-name` regardless of any other arguments, but only because it has a different naming convention that we don't plan to repeat for any future registration.

Reply:
@glasserm, can you confirm my guess or explain `--registration-name` and `--concat-reg-name`?

Reply:
--registration-name is the name generated by MSMAll
--concat-reg-name is the final name that includes the MSMAll and dedrifting registrations
MSMSulc would not be appropriate for either of these.

Reply:
@mharms what do you want the description of `--registration-name` and `--concat-reg-name` to be?  Do you just want it to also convey "not MSMSulc"?

Reply:
Making clear that they are "not MSMSulc" would be helpful, but having come this far, I think it would be helpful to make them generally clearer as well.  For example, `--concat-reg-name` is indicated as optional, but it isn't at all clear under what situations its inclusion is actually necessary (which IIUC is anytime you are doing a "resampling" and not the "dedrifting".  Is it the case that `--dedrift-reg-files` and `--concat-reg-name` are mutually exclusive, and one of the two is required?  If so, making that clear would be helpful.

Reply:
--registration-name and --concat-reg-name would always be used.  If they are the same, --dedrift-reg-files can be omitted.

Reply:
Actually, I think ConcatRegName should be mandatory, it is used with no conditional in the script.  The dedrift files are only used when RegName and ConcatRegName are not equal.

Reply:
Making that all clear would I think be very helpful.

Reply:
Oh, I see, when ConcatRegName isn't specified, it defaults to RegName.

Reply:
Maybe we should make it an error to specify --concat-reg-name without --dedrift-reg-files, so we can just treat the option as an output naming convention.

Reply:
Also, I think it would help if we made clear that it is called `--concat-reg-name` because it represents the concatenation of MSMAll and dedrifting operations...?  (Hence, if you don't need to dedrift, there is no inherent concatenation, and thus `--registration-name` and `--concat-reg-name` will be the same).  Esp. given that the script now contains other `-concat-` variables which are an entirely different sort of concatenation (across runs).

Reply:
That would be a lot more intuitive to me.  Currently it is unclear which options are specifying inputs, and which are just controlling output names...

Reply:
Here is what I would edit it to:

```
  [--dedrift-reg-files=</Path/to/File/Left.sphere.surf.gii@/Path/to/File/Right.sphere.surf.gii>]
       Path to the spheres output by the MSMRemoveGroupDrift pipeline. This option should not be used in
       most circumstances.
  [--concat-reg-name=<regname>] Output name for the dedrifted registration (referred to as the
       concatenated registration), requires using --dedrift-reg-files. This option should not be used in
       most circumstances.
```

Reply:
Ok, I'm back to somewhat confused.  Here is how we actually called DeDrift for an example HCA subject:
```
# Running: /opt/HCP/HCPpipelines/DeDriftAndResample/DeDriftAndResamplePipeline.sh             --path="/home/hodgem/XNAT_BUILD_DIR/hodgem/msmall_Q2_2020_runs/CCF_HCA_STG/MsmAllProcessing.HCA6347674_V1_MR.1589590114.XNAT_PROCESS_DATA/HCA6347674_V1_MR/subjects/HCA6347674_V1_MR/hcp"             --subject="HCA6347674_V1_MR"             --high-res-mesh="164"             --low-res-meshes="32"             --registration-name="MSMAll_InitialReg_2_d40_WRN"             --dedrift-reg-files="/opt/HCP/HCPpipelines/global/templates/MSMAll/DeDriftingGroup.L.sphere.DeDriftMSMAll.164k_fs_LR.surf.gii@/opt/HCP/HCPpipelines/global/templates/MSMAll/DeDriftingGroup.R.sphere.DeDriftMSMAll.164k_fs_LR.surf.gii"             --concat-reg-name="MSMAll"             --maps="sulc@curvature@corrThickness@thickness"             --myelin-maps="MyelinMap@SmoothedMyelinMap"             --multirun-fix-names="rfMRI_REST1_AP@rfMRI_REST1_PA@tfMRI_VISMOTOR_PA@tfMRI_CARIT_PA@tfMRI_FACENAME_PA@rfMRI_REST2_AP@rfMRI_REST2_PA"             --multirun-fix-concat-names="fMRI_CONCAT_ALL"             --fix-names="NONE"             --dont-fix-names="NONE"             --smoothing-fwhm="2"             --high-pass="0"             --matlab-run-mode="0"             --motion-regression="FALSE"             --myelin-target-file="NONE"             --input-reg-name="NONE"
```
Note that we used `--dedrift-reg-files`, which seems inconsistent with the proposed statement that "This option should not be used in most circumstances".

Reply:
`--registration-name="MSMAll_InitialReg_2_d40_WRN" --concat-reg-name="MSMAll"` hints at the difference.

Other users will use MSMAll to register directly to the final (already dedrifted) template.  I'm not sure what we did in the group that we generated the template from, but the technically correct thing to do would have been to take the final (already dedrifted) MSMAll template and run another fresh registration to it, just like other users will on their data.  Hopefully this was just a shortcut to save processing time that doesn't result in an appreciable registration difference?

Reply:
This is how we have run things since HCP-YA.  I don't think there is much effect of this for this registration.

Reply:
Oh wait, in an HCA subject?  That seems unfortunate, I don't think we should have done that...

Reply:
Well, if it uses the same dedrifting files as in any other group, and doesn't recompute drift, I guess that is okay, it mostly implies an extra shift that should have been in the template to begin with.  Using dedrifting options just to match an existing registration (rather than actually *doing* a dedrifting) makes these options much harder to explain.

Reply:
Assuming that we tell others to run things the way we do, here is a new stab at edited usage info:
```
   --registration-name=<regname> String corresponding to the input registration sphere name 
       (e.g. \${Subject}.\${Hemisphere}.sphere.\${RegName}.native.surf.gii), usually MSMAll_InitialReg...
  [--dedrift-reg-files=</Path/to/File/Left.sphere.surf.gii@/Path/to/File/Right.sphere.surf.gii>]
       Usually the spheres in global/templates/MSMAll/
  [--concat-reg-name=<regname>] String corresponding to the output name for the dedrifted registration
       (referred to as the concatenated registration), usually MSMAll. Requires --dedrift-reg-files
```

Reply:
So, it is still unclear to me, is the reason we are doing it this way because the templates in `global/templates/MSMAll` were not themselves dedrifted?

What remains unclear in the proposed text above is under what situation `--dedrift-reg-files` and `--concat-reg-name` can actually be omitted...

Reply:
I suspect that the template functional data was dedrifted, and that the "drift" files are from just a single iteration of re-register and dedrift.  I talked to Matt about how if other studies computed drifts from their registrations and dedrifted, that could make the alignment across studies worse, and how that means that you shouldn't try to dedrift the final iteration of registration out of subjects' registrations, but that may have been after it was already run on most of the YA cohort.

When/if we come up with a better surface registration, I think we should avoid this situation, such that we would then update these descriptions to say "when using MSMAll, do this, otherwise don't specify this option".  No sense in making it harder to understand when there is only one registration to use with this script.

Reply:
Yes, it would have made things much simpler if the dedrifting was applied only to the template, and then the registrations were run one more time to get the final MSMAll alignment, without applying any extra shift afterwards.  That way, users could just run MSMAll, resample using just that sphere (no dedrift files, no concatenation), and they are done.  Really, in my opinion, this script should never have said anything about dedrifting, as it isn't something that should be done outside of template generation.

As for user choice, if the dedrift files are from only the last iteration of template refinement, then the shifts in them should be small, and choosing not to use them would be largely imperceptible.  Unfortunately, since that isn't what we did for HCP YA, we are stuck with telling users to specify these extra arguments (because even a small systematic bias is a problem).

Reply:
I think a description of all this as part of a README would be very helpful for users attempting to understand dedrifting and how it relates to MSMAll.  I realize that said README might have to be a little bit messy, perhaps with both "optimal implementation" and "actual/current implementation" components, but that's the reality of the current situation.

Reply:
This has nothing to do with this pull request.  It would be fine to revisit the implementation when we overhaul MSMAll later this year.

Reply:
FWIW, I'm still not sure I fully understand the situation myself... IIUC, the issue is NOT that the MSMAll templates were not dedrifted, but that an extra (unnecessary) round of dedrifting was applied, and now we are "stuck" replicating the application of that extra dedrift round because otherwise a small bias might be present relative to what we've previously done?

Reply:
I think that is accurate, but Matt is the only one who knows for sure.  What he has said suggests that it is likely the case.

To be clear, in groups other than the HCP YA, it doesn't count as a round of dedrifting at all, because the "extra round" is derived from HCP YA data and not recomputed from the new group - this is intentional, to avoid causing a different group to add bias via a different dedrifting (in particular, this dedrifting is computed relative to MSMSulc, and therefore could be large if recomputed in a different population).

Reply:
Resolving this conversation by making it an independent issue: https://github.com/Washington-University/HCPpipelines/issues/185